File: Day76/Part5/achievement_chain.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

bool solve() {
    int n, m;
    if (!(cin >> n >> m)) return false;
    
    vector<vector<int>> adj(n);
    vector<int> in_degree(n, 0);
    
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        if (u >= 0 && u < n && v >= 0 && v < n) {
            adj[u].push_back(v);
            in_degree[v]++;
        }
    }
    
    vector<int> bonus(n);
    for (int i = 0; i < n; ++i) {
        cin >> bonus[i];
    }
    
    queue<int> q;
    for (int i = 0; i < n; ++i) {
        if (in_degree[i] == 0) {
            q.push(i);
        }
    }
    
    int count = 0;
    while (!q.empty()) {
        int u = q.front();
        q.pop();
        count++;
        
        // Process normal dependencies
        for (int v : adj[u]) {
            in_degree[v]--;
            if (in_degree[v] == 0) {
                q.push(v);
            }
        }
        
        // Process bonus
        int target = bonus[u];
        if (target >= 0 && target < n) {
            in_degree[target]--;
            if (in_degree[target] == 0) {
                q.push(target);
            }
        }
    }
    
    return count == n;
}

int main() {
    cout << (solve() ? "true" : "false") << endl;
    return 0;
}

File: Day76/Part5/achievement_chain.py
import sys
from collections import deque

def solve():
    try:
        input_data = sys.stdin.read().split()
        if not input_data:
            return
        
        iterator = iter(input_data)
        try:
            n = int(next(iterator))
            m = int(next(iterator))
        except StopIteration:
            return
            
        adj = [[] for _ in range(n)]
        in_degree = [0] * n
        
        for _ in range(m):
            u = int(next(iterator))
            v = int(next(iterator))
            if 0 <= u < n and 0 <= v < n:
                adj[u].append(v)
                in_degree[v] += 1
                
        bonus = []
        for _ in range(n):
            bonus.append(int(next(iterator)))
            
        q = deque()
        for i in range(n):
            if in_degree[i] == 0:
                q.append(i)
                
        count = 0
        while q:
            u = q.popleft()
            count += 1
            
            # Normal dependencies
            for v in adj[u]:
                in_degree[v] -= 1
                if in_degree[v] == 0:
                    q.append(v)
            
            # Bonus
            target = bonus[u]
            if 0 <= target < n:
                in_degree[target] -= 1
                if in_degree[target] == 0:
                    q.append(target)
                    
        if count == n:
            print("true")
        else:
            print("false")
            
    except Exception:
        pass

if __name__ == "__main__":
    solve()
